### PR TITLE
Disallow phantom type parameter/member in classes

### PR DESCRIPTION
**Pull request by nicolasstucki**
This fixes potential ways to fake evidences by casting an object that has a phantom type parameter/member. See `neg/phantom-class-type-parameters.scala` as an example.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala [hunk 1]
```diff
@@ -506,8 +506,16 @@ trait TypeAssigner {
...
+  def assignType(tree: untpd.TypeDef, sym: Symbol)(implicit ctx: Context) = {
+    val tdef = tree.withType(symbolicIfNeeded(sym).orElse(sym.typeRef))
+    val isParam = tdef.tpe.typeSymbol.is(Param)
+    val isPhantom = tdef.tpe.isPhantom
+    if (isParam && isPhantom && tdef.symbol.owner.isPrimaryConstructor)
+      ctx.error("Classes cannot have phantom type parameters", tdef.pos)
+    else if (!isParam && isPhantom && !tdef.symbol.owner.isStaticOwner)
+      ctx.error("Non static classes cannot have phantom type members", tdef.pos)
+    tdef
+  }
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala]: A small comment: may this be done in RefChecks where a lot of checking happens?

Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala]: I will try. I also felt uneasy adding all those checks in there.

Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala]: Moved it to Checking

**Comment by nicolasstucki** [comment]
Found a cleaner solution with #2525

**Pull request closed without merging**